BREAKING CHANGE: Refactor Tooltip component

imoutaharik opened the pull request:
*Simplify Tooltip component usage for developers
- Removed TooltipArrow
- Removed React-aria
- Popper moved inside the component
- Added positioning (position and alignment), flip, and arrow
- Added arrow styling
- Added timeout on hover
- Accessibility

lenardpalko commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +56,221 @@ const paddingsBottom = {
+                triggerRefElement.current.removeEventListener("click", () => setIsOpen(!isOpen));
+                triggerRefElement.current.removeEventListener("mouseover", () => setIsOpen(true));
+                triggerRefElement.current.removeEventListener("mouseleave", () =>
+                    setTimeout(() => setIsOpen(false), 2000),
```
Maybe we could add a config for the timeout value as well, that could have a default value if not set explicitly.

lenardpalko commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +56,221 @@ const paddingsBottom = {
+        if (triggerRefElement) {
+            triggerRefElement.current.addEventListener("click", () => setIsOpen(!isOpen));
+            triggerRefElement.current.addEventListener("mouseover", () => setIsOpen(true));
+            triggerRefElement.current.addEventListener("mouseleave", () => setTimeout(() => setIsOpen(false), 2000));
```
We either need to clear the timeout if the mouse is on the tooltip, either add a new variable e.g. `mouseOverTooltip` that would be true while the user is over the tooltip and which conditions the display of the tooltip as well. So the condition to display the tooltip would be then: `isOpen && mouseOverTooltip`. Otherwise the tooltip closes even if the user is on over the tooltip with the mouse.

lenardpalko commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +56,221 @@ const paddingsBottom = {
-                {...tooltipProps}
-            >
-                {brightHeader && <BrightHeader headerStyle={brightHeader} />}
+        <div>
```
Maybe we could directly use ReactPortal here to avoid any positioning issues with other elements on the page.

Kenny806 replied:
Agreed. Both me and Lenny currently have to use Portal to correctly position the Tooltip, would be great if this was built into the component itself.

imoutaharik replied:
@lenardpalko @Kenny806 I just updated according to all comments. I agree it makes sense to use Portal within the component. I also fixed the timeout issue.

Kenny806 replied:
@imoutaharik From my side it works fine now. I won't approve the PR since @thisguyscode still has some comments, but the way Tooltip is used is good for me and I have no further comments there.

imoutaharik commented:
@SamuelAlev I added you as a reviewer as I made a lot of changes in the tooltip component that may affect your team. 
@thisguyscode I also added you so you're aware of all the changes, I actually got rid of the tooltipArrow component that was passed as a child and used the logic that was thought for the popover.

thisguyscode commented:
@imoutaharik A few comments about the hover delay:

- There seem to be multiple properties in Storybook to control this, and I can't immediately see the effect of either.
- The tooltip should never be destroyed when the mouse is over it, regardless of hover delay
- In Storybook the delay is really high. I believe the default behavior should be almost 0 delay. Allowing just enough time to mouseover for the interactive tooltips.

thisguyscode commented:
@imoutaharik General Storybook:

- Could you add a story with multiple tooltips please? Would be cool to test how they behave when switching between
- I don't think you always need to add all properties to all stories. This way it's hard to find the relevant ones for the particular story.
- The `hoverDelay` / `hoverdelay` duplicate properties mentioned above.

lenardpalko commented:
@imoutaharik for my use case also seems to be all good. It simplifies a lot on how the tooltip should be used. Thanks for the great work 👍

SamCreasey commented on src/components/Tooltip/Tooltip.stories.tsx:
```diff
@@ -43,72 +43,128 @@ export default {
+            <button ref={firstTriggerRef} className="tw-mr-1">
+                <IconAttentionFilled size={IconSize.Size16} />
+            </button>
+            <Tooltip {...args} triggerRefElement={firstTriggerRef} withArrow={true} />
```
WithArrow Storybook control is overwritten by this prop and it feels a bit weird that it toggles one and not the other

SamCreasey commented:
If you hover over the tooltip itself it disappears, which is not great for when you have a link inside the tooltip

https://user-images.githubusercontent.com/93908356/166215714-b5b0d6f2-af92-4f38-84aa-7b0e8ea236f6.mov

SamCreasey commented:
> If you hover over the tooltip itself it disappears, which is not great for when you have a link inside the tooltip
> 
>  Screen.Recording.2022-05-02.at.11.44.54.mov

Also it is not possible to tab onto a link inside the tooltip

SamCreasey commented:
<img width="143" alt="Screenshot 2022-05-02 at 11 48 44" src="https://user-images.githubusercontent.com/93908356/166216001-8a17cbae-67b6-4aae-8fab-cd5928418f7e.png">
I think the focus style of the element when using keyboard navigation might need to be updated to match the exisiting focus styles of other components

SamCreasey commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +58,236 @@ const paddingsBottom = {
+    const arrowStyling = setArrowClasses();
+    const [isOpen, setIsOpen] = useState(false);
+
+    let timeout: ReturnType<typeof setTimeout> | null = null;
```
I wonder if timeout should be a ref object so that it persists between renders 🤔

SamCreasey commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -4,31 +4,31 @@ import { Button, ButtonSize, ButtonStyle } from "@components/Button/Button";
-type PopperAttributes = { [key: string]: string };
-
 export type TooltipProps = {
+    triggerRefElement?: MutableRefObject<any>;
```
Might consider changing any to HTMLElement, wdyt?

SamCreasey commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +58,236 @@ const paddingsBottom = {
+
+    const tooltipOffset = withArrow ? 10 : 5;
+    const popperInstance = usePopper(triggerRefElement?.current, tooltipContainerRef.current, {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        placement: placement,
```
This can be removed by casting the type of the placement to Placement

imoutaharik replied:
Fixed using variation Placement

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,268 @@ const paddingsBottom = {
+    ["right-End"]: "right-end",
+};
+
+export const Tooltip: FC<TooltipProps> = ({
```
[`FC` is not allowed anymore](https://frontify.slack.com/archives/C01JE9QENCF/p1650544457958619)

imoutaharik replied:
True! And I voted to keep it 😁 . I removed it. Thanks for all the feedback. I took into account all your comments and I just simplified everything again.

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,268 @@ const paddingsBottom = {
     const { linkProps } = useLink({}, linkRef);
-
     const { isFocusVisible, focusProps } = useFocusRing();
+    const hasLargePaddingTop = linkUrl || brightHeader || buttons || heading || headingIcon;
```
`useMemo` so it doesn't get recomputed for each render in case any other props changes

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,268 @@ const paddingsBottom = {
-    const hasLargePaddingTop = linkUrl || brightHeader || buttons || heading || headingIcon;
+    const setArrowClasses = () => {
+        const currentPlacement = popperInstance.state?.placement ?? position;
+        switch (true) {
+            case currentPlacement.toString().includes(TooltipPosition.Top):
+                return "before:tw-border-t-0 before:tw-border-l-0 tw-bottom-[-6px] before:tw-bg-black-100 before:dark:tw-bg-white";
+            case currentPlacement.toString().includes(TooltipPosition.Right):
+                return merge([
+                    "before:tw-border-t-0 before:tw-border-r-0 tw-left-[-6px]",
+                    brightHeader && alignment === TooltipAlignment.Start
+                        ? brightHeaderArrowBackgroundColors[brightHeader]
+                        : "before:tw-bg-black-100 before:dark:tw-bg-white",
+                ]);
+            case currentPlacement.toString().includes(TooltipPosition.Bottom):
+                return merge([
+                    "before:tw-border-b-0 before:tw-border-r-0 tw-top-[-6px]",
+                    brightHeader
+                        ? brightHeaderArrowBackgroundColors[brightHeader]
+                        : "before:tw-bg-black-100 before:dark:tw-bg-white",
+                ]);
+            case currentPlacement.toString().includes(TooltipPosition.Left):
+                return merge([
+                    "before:tw-border-b-0 before:tw-border-l-0 tw-right-[-6px]",
+                    brightHeader && alignment === TooltipAlignment.Start
+                        ? brightHeaderArrowBackgroundColors[brightHeader]
+                        : "before:tw-bg-black-100 before:dark:tw-bg-white",
+                ]);
+            default:
+                return "before:tw-border-b-0 before:tw-border-r-0 tw-top-[-6px] before:tw-bg-black-100 before:dark:tw-bg-white";
```
Can we make it as a `const` outside the component?
If not, we can at least wrap it in a `useCallback`

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,268 @@ const paddingsBottom = {
+        setIsOpen(true);
+    };
+
+    const handleHideTooltipOnHover = () => {
+        timeoutRef.current = setTimeout(() => setIsOpen(false), hoverDelay);
+    };
```
`useCallback`

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,268 @@ const paddingsBottom = {
+        timeoutRef.current = setTimeout(() => setIsOpen(false), hoverDelay);
+    };
+
+    const checkIfHovered = (e: Event) => {
```
```suggestion
    const checkIfHovered = (event: Event) => {
```

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,268 @@ const paddingsBottom = {
+    const firstButtonRef = useRef<HTMLButtonElement | null>(null);
+    const lastButtonRef = useRef<HTMLButtonElement | null>(null);
+
+    useEffect(() => {
```
Better to add the events as props of `TooltipIcon` and use the `on...` notation in the component instead of vanilla events and ref

SamCreasey commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,263 @@ const paddingsBottom = {
-                            data-test-id="tooltip-link"
-                            ref={linkRef}
-                            href={linkUrl}
-                            target="_blank"
-                            rel="noopener noreferrer"
-                            className={merge([
-                                "tw-text-xs tw-text-black-40 dark:tw-text-black-80 tw-underline tw-mt-1",
-                                isFocusVisible && FOCUS_STYLE,
-                            ])}
+                {triggerElement &&
+                    cloneElement(triggerElement, {
+                        ref: triggerRefElement,
+                    })}
+            </div>
```
Wrapping the trigger element in a div brings about the same issues that we had with the FlyoutTrigger, where you might have a component that directly relies on it's parent for styling and cannot anymore. I would maybe suggest a render function.
`const triggerProps = {onMouseOver, onMouseLeave, onFocus, onBlur}`
```
<>
  {triggerElement(triggerProps, triggerRef)}
</>
```
wdyt?

imoutaharik replied:
Will create a task for it.

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,263 @@ const paddingsBottom = {
+    End = "End",
+}
+
+const placementMap: Record<string, VariationPlacement> = {
+    ["top-Start"]: "top-start",
+    ["top-End"]: "top-end",
+    ["bottom-Start"]: "bottom-start",
+    ["bottom-End"]: "bottom-end",
+    ["left-Start"]: "left-start",
+    ["left-End"]: "left-end",
+    ["right-Start"]: "right-start",
+    ["right-End"]: "right-end",
+};
```
```suggestion
const placementMap: Record<`${TooltipPosition}-${TooltipAlignment}`, VariationPlacement> = {
    ["top-Start"]: "top-start",
    ["top-End"]: "top-end",
    ["bottom-Start"]: "bottom-start",
    ["bottom-End"]: "bottom-end",
    ["left-Start"]: "left-start",
    ["left-End"]: "left-end",
    ["right-Start"]: "right-start",
    ["right-End"]: "right-end",
};
```
And seems we have different casing

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -72,36 +72,40 @@ export enum TooltipAlignment {
-    ["right-Start"]: "right-start",
-    ["right-End"]: "right-end",
+const placementMap: Record<`${TooltipPosition}-${TooltipAlignment}`, Placement> = {
+    ["Top-Start"]: "top-start",
```
Can be written "[`${TooltipPosition.Top}-${TooltipAlignment.Start}`]"

imoutaharik replied:
Typescript didn't like that at all.

SamuelAlev commented on src/components/Tooltip/Tooltip.tsx:
```diff
@@ -56,113 +59,267 @@ const paddingsBottom = {
+
+const getArrowClasses = (currentPlacement: string, brightHeader: BrightHeaderStyle | undefined, alignment: string) => {
+    switch (true) {
+        case currentPlacement.toString().includes("top"):
```
Isn't it 
```suggestion
        case currentPlacement.toString().includes(TooltipPosition.Top):
```

imoutaharik replied:
It's TooltipPosition.Top.toLowerCase(), Do we want it like this?

SamuelAlev replied:
I read the code a bit more, now I see why 👍
